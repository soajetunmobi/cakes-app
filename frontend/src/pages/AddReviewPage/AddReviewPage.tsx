import React, { FormEvent, FunctionComponent, useState } from 'react'
import { useHistory } from 'react-router-dom'
import { routes } from '../../common/const/routes'
import { cakeService } from '../../service'
import { ErrorPanel } from '../../components/ErrorPanel'

export const AddReviewPage: FunctionComponent = () => {
  const history = useHistory()

  const [review, setReview] = useState({
    name: '',
    imageUrl: '',
    comment: '',
    yumFactor: 0
  })
  const [error, setError] = useState<string | undefined>(undefined)

  function handleChange(e: FormEvent<HTMLInputElement>) {
    const { name, value } = e.currentTarget
    setReview(review => ({ ...review, [name]: value }))
  }

  function handleCommentChange(e: FormEvent<HTMLTextAreaElement>) {
    const { name, value } = e.currentTarget

    if (value.length >= 5 || value.length <= 200) {
      setReview(review => ({ ...review, [name]: value }))
    }
  }

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault()

    if (isNameValid() && isCommentValid() && isYumFactorValid() && isImageUrlValid()) {
      try {
        await cakeService.add(review)
        if (!error) {
          history.push(routes.cakes)
        }
      } catch (err) {
        console.log('error', err)
        setError('There was an error saving your data')
      }
    }
  }

  const isYumFactorValid = (): boolean => {
    return review.yumFactor >= 1 && review.yumFactor <= 5
  }

  const saveYumFactorValue = (yumFactor: number) => {
    setReview(review => ({ ...review, yumFactor: yumFactor }))
  }

  const goBackToCakeList = (): void => {
    history.push(routes.cakes)
  }

  function isNameValid(): boolean {
    return review.name.length > 3
  }

  function validateUrl(url: string): boolean {
    return /((ftp|https?):\/\/)?(www\.)?[a-z0-9\-\.]{3,}\.[a-z]{2,3}$/.test(url)
  }

  function isImageUrlValid(): boolean {
    return validateUrl(review.imageUrl) && review.imageUrl.length > 1
  }

  function isCommentValid(): boolean {
    return review.comment.length >= 5 && review.comment.length <= 200
  }

  return (
    <div className='form-wrapper'>
      {error && (
        <ErrorPanel>
          <div>{error}</div>
        </ErrorPanel>
      )}

      <h2 className='page__title'>Add comment</h2>
      <form name='form' className='form' onSubmit={handleSubmit}>
        <div className='form__group'>
          <label>Name</label>
          <input
            type='text'
            name='name'
            data-testid='review-name'
            value={review.name}
            required
            onChange={handleChange}
            onBlur={isNameValid}
            className={'form__input'}
          />
          {!isNameValid() && (
            <div data-testid='name-error' className='invalid-feedback'>
              Cake name is required
            </div>
          )}
        </div>
        <div className='form__group'>
          <label>Image url</label>
          <input
            type='text'
            name='imageUrl'
            data-testid='review-image'
            value={review.imageUrl}
            required
            onChange={handleChange}
            onBlur={isImageUrlValid}
            className={'form__input'}
          />
          {!isImageUrlValid() && (
            <div data-testid='image-error' className='invalid-feedback'>
              Image url is required
            </div>
          )}
        </div>
        <div className='form__group'>
          <label>Comment</label>
          <textarea
            rows={2}
            data-testid='review-comment'
            maxLength={200}
            required
            name='comment'
            value={review.comment}
            onChange={handleCommentChange}
            onBlur={isCommentValid}
            className={'form__textarea'}
          />
          {!isCommentValid() && (
            <div data-testid='comment-error' className='invalid-feedback'>
              Comment is required (At least 5 characters and at most 200 characters)
            </div>
          )}
        </div>
        <div className='form__group no-border'>
          <label>Yum factor</label>
          <div>
            <span
              id='1'
              data-testid='yum-1'
              className={`circle cursor--pointer${review.yumFactor === 1 ? ' active' : ''}`}
              onClick={() => saveYumFactorValue(1)}
            >
              1
            </span>
            <span
              id='2'
              data-testid='yum-2'
              className={`circle cursor--pointer${review.yumFactor === 2 ? ' active' : ''}`}
              onClick={() => saveYumFactorValue(2)}
            >
              2
            </span>
            <span
              id='3'
              data-testid='yum-3'
              className={`circle cursor--pointer${review.yumFactor === 3 ? ' active' : ''}`}
              onClick={() => saveYumFactorValue(3)}
            >
              3
            </span>
            <span
              id='4'
              data-testid='yum-4'
              className={`circle cursor--pointer${review.yumFactor === 4 ? ' active' : ''}`}
              onClick={() => saveYumFactorValue(4)}
            >
              4
            </span>
            <span
              id='5'
              data-testid='yum-5'
              className={`circle cursor--pointer${review.yumFactor === 5 ? ' active' : ''}`}
              onClick={() => saveYumFactorValue(5)}
            >
              5
            </span>
          </div>
          {!isYumFactorValid() && (
            <div data-testid='error-yumfactor' className='invalid-feedback'>
              Yum factor is required
            </div>
          )}
        </div>
        <div className='buttons-container buttons-container--space-between'>
          <button onClick={() => goBackToCakeList()} className='button--secondary'>
            Cancel
          </button>
          <button className='button--primary'>Submit</button>
        </div>
      </form>
    </div>
  )
}
