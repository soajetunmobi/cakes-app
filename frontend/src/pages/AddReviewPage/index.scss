

.form {
   width: 100%;

  &__group {
  width: 100%;
  position: relative;
  border-bottom: 2px solid #adadad;
  margin-bottom: 37px;
}

  &__input {
    display: block;
    width: 100%;
    font-size: px-to-rem(15px);
    color: #555;
    line-height: 1.2;
    border: none;
    outline: none;
    height: px-to-rem(45px);
  }

  &__textarea {
    min-height: px-to-rem(80px);
    padding-top: 13px;
    padding-bottom: 13px;
    display: block;
    width: 100%;
    font-size: px-to-rem(15px);
    color: #555;
    line-height: 1.2;
    border: none;
    outline: none;
  }
}
