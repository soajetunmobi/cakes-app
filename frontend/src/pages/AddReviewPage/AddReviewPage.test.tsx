import React from 'react'
import renderer from 'react-test-renderer'
import { AddReviewPage } from './AddReviewPage'
import { render, fireEvent } from '@testing-library/react'

describe('Add review component', () => {
  test('it renders correctly', () => {
    const root = renderer.create(<AddReviewPage />).toJSON()
    expect(root).toMatchSnapshot()
  })

  describe('Add review form is rendered', () => {
    it('should render form', async () => {
      const { getByText } = render(<AddReviewPage />)
      const nameInput = getByText('Name')
      expect(nameInput).toBeInTheDocument()
      const imageInput = getByText('Image url')
      expect(imageInput).toBeInTheDocument()
      const commentInput = getByText('Comment')
      expect(commentInput).toBeInTheDocument()
      const yumFactorContainer = getByText('Yum factor')
      expect(yumFactorContainer).toBeInTheDocument()
    })

    it('should show error messages in form', async () => {
      const { getByTestId } = render(<AddReviewPage />)
      const nameError = getByTestId('name-error')
      expect(nameError).toBeInTheDocument()
      const imageError = getByTestId('image-error')
      expect(imageError).toBeInTheDocument()
      const commentError = getByTestId('comment-error')
      expect(commentError).toBeInTheDocument()
      const yumError = getByTestId('error-yumfactor')
      expect(yumError).toBeInTheDocument()
    })

    it('should not show error messages in form when data is entered', async () => {
      const { getByTestId, queryByTestId } = render(<AddReviewPage />)
      const nameInput = getByTestId('review-name')
      fireEvent.change(nameInput, { target: { value: 'Cake' } })
      expect(queryByTestId('/name-error/i')).toBeNull()

      const imageInput = getByTestId('review-image')
      fireEvent.change(imageInput, { target: { value: 'www.test.com' } })
      expect(queryByTestId('/image-error/i')).toBeNull()

      const commentInput = getByTestId('review-comment')
      fireEvent.change(commentInput, { target: { value: 'This is comment' } })
      expect(queryByTestId('/comment-error/i')).toBeNull()

      const yum3 = getByTestId('yum-3')
      fireEvent.click(yum3)
      expect(queryByTestId('/error-yumfactor/i')).toBeNull()
    })
  })
})
