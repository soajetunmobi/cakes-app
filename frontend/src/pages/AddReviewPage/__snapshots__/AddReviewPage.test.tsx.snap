// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add review component it renders correctly 1`] = `
<div
  className="form-wrapper"
>
  <h2
    className="page__title"
  >
    Add comment
  </h2>
  <form
    className="form"
    name="form"
    onSubmit={[Function]}
  >
    <div
      className="form__group"
    >
      <label>
        Name
      </label>
      <input
        className="form__input"
        data-testid="review-name"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
      <div
        className="invalid-feedback"
        data-testid="name-error"
      >
        Cake name is required
      </div>
    </div>
    <div
      className="form__group"
    >
      <label>
        Image url
      </label>
      <input
        className="form__input"
        data-testid="review-image"
        name="imageUrl"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
      <div
        className="invalid-feedback"
        data-testid="image-error"
      >
        Image url is required
      </div>
    </div>
    <div
      className="form__group"
    >
      <label>
        Comment
      </label>
      <textarea
        className="form__textarea"
        data-testid="review-comment"
        maxLength={200}
        name="comment"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        rows={2}
        value=""
      />
      <div
        className="invalid-feedback"
        data-testid="comment-error"
      >
        Comment is required (At least 5 characters and at most 200 characters)
      </div>
    </div>
    <div
      className="form__group no-border"
    >
      <label>
        Yum factor
      </label>
      <div>
        <span
          className="circle cursor--pointer"
          data-testid="yum-1"
          id="1"
          onClick={[Function]}
        >
          1
        </span>
        <span
          className="circle cursor--pointer"
          data-testid="yum-2"
          id="2"
          onClick={[Function]}
        >
          2
        </span>
        <span
          className="circle cursor--pointer"
          data-testid="yum-3"
          id="3"
          onClick={[Function]}
        >
          3
        </span>
        <span
          className="circle cursor--pointer"
          data-testid="yum-4"
          id="4"
          onClick={[Function]}
        >
          4
        </span>
        <span
          className="circle cursor--pointer"
          data-testid="yum-5"
          id="5"
          onClick={[Function]}
        >
          5
        </span>
      </div>
      <div
        className="invalid-feedback"
        data-testid="error-yumfactor"
      >
        Yum factor is required
      </div>
    </div>
    <div
      className="buttons-container buttons-container--space-between"
    >
      <button
        className="button--secondary"
        onClick={[Function]}
      >
        Cancel
      </button>
      <button
        className="button--primary"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
